com.ca.lsp.cobol/lsp-service-cobol/src/test/java/com/ca/lsp/cobol/usecases/TestCopybookWithIndirectRecursiveDependencyIsDetected.java

```diff
@@ -15,54 +15,63 @@
 package com.ca.lsp.cobol.usecases;
 
 import com.ca.lsp.cobol.positive.CobolText;
+import com.ca.lsp.cobol.usecases.engine.UseCaseEngine;
 import org.eclipse.lsp4j.Diagnostic;
-import org.eclipse.lsp4j.Range;
 import org.junit.Test;
 
-import java.util.List;
+import java.util.HashMap;
+import java.util.Map;
 
-import static org.junit.Assert.assertEquals;
+import static com.ca.lsp.cobol.service.delegates.validations.SourceInfoLevels.ERROR;
+import static com.ca.lsp.cobol.service.delegates.validations.SourceInfoLevels.INFO;
+import static java.util.Arrays.asList;
+import static org.eclipse.lsp4j.DiagnosticSeverity.Error;
+import static org.eclipse.lsp4j.DiagnosticSeverity.Information;
 
 /**
- * This test checks the error shown if the copybook that is used in the Cobol file contains a link
+ * This test checks the error shown if the copybook that is used in the COBOL file contains a link
  * to another one, that has a dependency to the first copybook.
  */
-public class TestCopybookWithIndirectRecursiveDependencyIsDetected extends NegativeUseCase {
+public class TestCopybookWithIndirectRecursiveDependencyIsDetected {
 
   private static final String TEXT =
-      "        IDENTIFICATION DIVISION.\r\n"
-          + "        PROGRAM-ID. test1.\r\n"
-          + "        DATA DIVISION.\r\n"
-          + "        WORKING-STORAGE SECTION.\r\n"
-          + "        COPY INDIRECT-COPY.\n\n"
-          + "        PROCEDURE DIVISION.\n\n";
+      "        IDENTIFICATION DIVISION.\n"
+          + "        PROGRAM-ID. test1.\n"
+          + "        DATA DIVISION.\n"
+          + "        WORKING-STORAGE SECTION.\n"
+          + "        COPY {~INDIRECT-COPY|1|3}.\n"
+          + "        PROCEDURE DIVISION.\n";
 
-  private static final String INDIRECT_COPY = "        COPY INNER-COPY.";
-  private static final String INNER_COPY = "        COPY INDIRECT-COPY.";
+  private static final String INDIRECT = "        COPY {~INNER-COPY|2|4}.";
+  private static final String INNER_COPY = "        COPY {~INDIRECT-COPY|3}.";
 
-  public TestCopybookWithIndirectRecursiveDependencyIsDetected() {
-    super(TEXT);
-  }
+  private static final String INDIRECT_NAME = "INDIRECT-COPY";
+  private static final String INNER_COPY_NAME = "INNER-COPY";
+
+  private static final String MESSAGE_RECURSION = "Recursive copybook declaration for: ";
+  private static final String MESSAGE_LONG_DECLARATION =
+      "Copybook declaration has more than 8 characters for: ";
 
-  @Override
   @Test
   public void test() {
-    super.test(
-        List.of(
-            new CobolText("INNER-COPY", INNER_COPY),
-            new CobolText("INDIRECT-COPY", INDIRECT_COPY)));
-  }
+    Map<String, Diagnostic> expectedDiagnostics = new HashMap<>();
+    expectedDiagnostics.put(
+        "1", new Diagnostic(null, MESSAGE_RECURSION + INDIRECT_NAME, Error, ERROR.getText()));
```

Comment:
in lsp4j  Diagnostic constructor defined with @NonNull, is that validation strict enough or ignored?  Same for all constructor where "null" used.

Reply:
If I understand correctly, this validation applies only while serialization to JSON.
Technically, this usage is not correct, but this check won't be applied in this scenario.